QueryGenerator.php

```diff
@@ -449,6 +449,33 @@ private function getGlue($method) {
          return ") OR (";
       }
    }
+
+   /**
+    * Debug method to print out the full query with the params inserted at the correct locations
+    */
+   public function toDebugString(): string {
+      [$query, $params] = $this->build();
+      $chunks = array_chunk($params, 2);
+      $paramValues = array_map(function ($chunk) {
+         [$type, $value] = $chunk;
+         return match ($type) {
+            'i' => (int)$value,   // Integer / boolean
+            'd' => (float)$value, // Float / Double
+            's' => "'$value'",    // String / Character
+            'j' => "'".addslashes(json_encode($value))."'", // Json
+            'e' => match (true) { // Enum
+                  $value instanceof Enum => $value->getValue(),
+                  $value instanceof BackedEnum => $value->value,
```

Comment:
Neither of these classes exist in this repo. Encoding parameters is tricky to do correctly. I know this is meant to be a debug output, but if it's close enough to the real query, it may be used that way. Here, we're doing it unsafely (for string parameters).

Given that we have a function that does this in the DB class (and is used to stringify a query so it can be safely ran), I'd vote for using that.


```php
Log::debug($db->queryToDebugString($q));
```

Reply:
I still think its not an easy interface to need to use the DB decorator, but it seems like thats the only way to get this to work in a cross-repo safe manner.

Reply:
Hmm..  maybe I'm missing something about the intended use.

I figure that everywhere we are constructing queries we are likely already executing them, we'd just like to *log* the query for debugging purposes in the process.

```php
$q = $db->query()->select(...)->where(...);
\Log::debug($q->toDebugString()); // this pull
\Log::debug($db->queryToDebugString($q)); // my proposition
$rows = $db->getAll($q);
\Log::debug($db->printLast(true)); // Already works with no changes if you're ok with running the query first
```

Reply:
That is the intended use, but its a weird concept to require another class to be able to show me what is in the QueryGenerator object, especially when the QuerySelector doesn't care about the specific DB its being run with. 

My big gripe with having it in the DB.php class is discoverability. I've looked for a function like this or `DB::printLast()` many times in the past but wasn't able to find it because I would never expect it to be outside of the objects own class (i.e. QueryGenerator).

Reply:
I still think its worth implementing a `queryToDebugString()` function in DB.php because sometimes we can't run a query before being able to `printLast()` if there are syntax errors (which is a big reason we would want to debug print a query).

Reply:
I'm on the fence here, but I think part of the disconnect is that by design `QueryGenerator` doesn't want to know how you handle parameters. You could use it to generate SQL for a totally different DB connector than the one that we happen to use it with. So generating a runnable query isn't _really_ in its domain: its domain is building a query and associated parameters. The `build()` function already gives you the closest thing to a complete query that this class is supposed to know how to generate.

That said, I _do_ think we're missing a function on `DB` like the one proposed here that lets you print out a query with parameters replaced _before_ running it. I don't see a reason not to add that.

Reply:
Successor pull: https://github.com/iFixit/ifixit/pull/50515

Reply:
> sometimes we can't run a query before being able to printLast() if there are syntax errors (which is a big reason we would want to debug print a query).

Good point @aburke07 !

> but its a weird concept to require another class to be able to show me what is in the QueryGenerator object,

I understand, but I think that it already does this with `->build()`. @sctice-ifixit did a better job of communicating my point than I did :-) :point_down: 

> The build() function already gives you the closest thing to a complete query that this class is supposed to know how to generate.